extensions/amp-tiktok/0.1/amp-tiktok.js

```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CSS} from '../../../build/amp-tiktok-0.1.css';
+import {Deferred} from '../../../src/utils/promise';
+import {Services} from '../../../src/services';
+import {
+  childElementByTag,
+  createElementWithAttributes,
+  removeElement,
+} from '../../../src/dom';
+import {debounce} from '../../../src/utils/rate-limit';
+import {getData, listen} from '../../../src/event-helper';
+import {htmlFor} from '../../../src/static-template';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {px, resetStyles, setStyles} from '../../../src/style';
+import {tryParseJson} from '../../../src/json';
+
+let id = 0;
+export class AmpTiktok extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {!Element} */
+    this.iframe_ = null;
+
+    /** @private {string} */
+    this.videoId_ = null;
+
+    /** @private {?Function}*/
+    this.unlistenMessage_ = null;
+
+    /** @private {?Promise} */
+    this.resolveReceivedFirstMessage_ = null;
+
+    /** @private {string} */
+    this.iframeNameString_ = this.getIframeNameString_();
+
+    /**
+     * @private {number}
+     * This value is calculated by multiplying our fixed width (325px)
+     * by the video aspect ratio (13:23 or 1.77) and adding 200px to account
+     * for the height of the caption.
+     */
+    this.fallbackHeight_ = 775;
+
+    this.resizeOuter_ = (height) => {
+      resetStyles(this.iframe_, [
+        'width',
+        'height',
+        'position',
+        'opacity',
+        'pointer-events',
+      ]);
+      this.iframe_.removeAttribute('aria-hidden');
+      this.iframe_.setAttribute('aria-label', 'Tiktok');
+      this.iframe_.classList.remove('i-amphtml-tiktok-unresolved');
+      this.iframe_.classList.add('i-amphtml-tiktok-centered');
+      this.forceChangeHeight(height);
+    };
+
+    this.resizeOuterDebounced_ = debounce(
+      this.win,
+      (height) => {
+        this.resizeOuter_(height);
+      },
+      1000
```

Comment:
FMI - is there a reason 1s was picked?

Reply:
The earliest TikTok messages all come in under 1s intervals so by debouncing for 1s we can be fairly certain that once 1s has passed we have received the final message